aten/src/ATen/native/cuda/AdaptiveAveragePooling3d.cu

```diff
@@ -0,0 +1,521 @@
+      output_size.size() == 1 || output_size.size() == 3,
+      "adaptive_avg_pool3d: internal error: output_size.size() must be 1 or 3");
+
+  int64_t osizeT = output_size[0];
+  int64_t osizeH = output_size[1];
+  int64_t osizeW = output_size[2];
+
+  int64_t sizeD, isizeT, isizeH, isizeW;
+  int64_t istrideD, istrideT, istrideH, istrideW;
+  int64_t totalZ;
+
+  const Tensor& input = input_.ndimension() == 4 ? input_ : input_.contiguous();
+
+  if (input.ndimension() == 4) {
+    sizeD = input.size(0);
+    isizeT = input.size(1);
+    isizeH = input.size(2);
+    isizeW = input.size(3);
+
+    istrideD = input.stride(0);
+    istrideT = input.stride(1);
+    istrideH = input.stride(2);
+    istrideW = input.stride(3);
+
+    output.resize_({sizeD, osizeT, osizeH, osizeW});
+
+    totalZ = sizeD * osizeT;
+  } else {
+    int64_t sizeB = input.size(0);
+    sizeD = input.size(1);
+    isizeT = input.size(2);
+    isizeH = input.size(3);
+    isizeW = input.size(4);
+
+    istrideD = input.stride(1);
+    istrideT = input.stride(2);
+    istrideH = input.stride(3);
+    istrideW = input.stride(4);
+
+    output.resize_({sizeB, sizeD, osizeT, osizeH, osizeW});
+
+    totalZ = sizeB * sizeD * osizeT;
+  }
+
+  AT_DISPATCH_FLOATING_TYPES_AND_HALF(
+      input.scalar_type(), "adaptive_avg_pool3d_cuda", [&] {
+        scalar_t* input_data = input.data<scalar_t>();
+        scalar_t* output_data = output.data<scalar_t>();
+
+        adaptiveaveragepool_loop(
+            input_data, output_data,
+            totalZ,
+            isizeT, isizeH, isizeW,
+            osizeT, osizeH, osizeW,
+            istrideD, istrideT, istrideH, istrideW);
+      });
+}
+
+void adaptive_avg_pool3d_backward_out_cuda_template(
+    Tensor& gradInput,
+    const Tensor& gradOutput_,
+    const Tensor& input) {
+  TensorArg grad_input_arg{gradInput, "gradInput", 1};
+  TensorArg grad_output_arg{gradOutput_, "gradOutput_", 2};
+  TensorArg input_arg{input, "input", 3};
+
+  checkAllSameGPU(
+      "adaptive_avg_pool3d_out_cuda",
+      {grad_input_arg, grad_output_arg, input_arg});
+
+  const Tensor gradOutput = gradOutput_.contiguous();
+
+  gradInput.resize_as_(input);
+  gradInput.zero_();
+
+  int64_t sizeD, isizeT, isizeH, isizeW;
+  int64_t osizeT, osizeH, osizeW;
+  int64_t totalZ;
+
+  if (input.ndimension() == 4) {
+    sizeD = input.size(0);
+    isizeT = input.size(1);
+    isizeH = input.size(2);
+    isizeW = input.size(3);
+
+    osizeT = gradOutput.size(1);
+    osizeH = gradOutput.size(2);
+    osizeW = gradOutput.size(3);
+  } else {
+    sizeD = input.size(1);
+    isizeT = input.size(2);
+    isizeH = input.size(3);
+    isizeW = input.size(4);
+
+    osizeT = gradOutput.size(2);
+    osizeH = gradOutput.size(3);
+    osizeW = gradOutput.size(4);
+  }
+
+  // somehow nonatomic is passing all test for volumetric case. (copied from
+  // THCUNN source code.)
+  // bool atomic = (isizeW%osizeW != 0) || (isizeH%osizeH != 0) || // (isizeT%osizeT != 0)
```

Comment:
Can you try re-enabling line 448?

Reply:
I know you took it over from `THCUNN`, but it's enabled in `AdaptiveMaxPool3d` for instance.

Reply:
I tried and it worked. So I chose line 448.